go/tools/builders/filter.go

```diff
@@ -37,7 +44,57 @@ func filterFiles(bctx build.Context, inputs []string) ([]string, error) {
 // it should be compiled.
 // TODO(#70): cross compilation: support GOOS, GOARCH that are different
 // from the host platform.
-func matchFile(bctx build.Context, input string) (bool, error) {
+func matchFile(bctx build.Context, input string, needPackage bool) (bool, bool, string, error) {
 	dir, base := filepath.Split(input)
-	return bctx.MatchFile(dir, base)
+	// First check tag filtering
+	match, err := bctx.MatchFile(dir, base)
+	if err != nil {
+		return false, false, "", err
+	}
+	// if we don't need the package, and we are cgo, no need to parse the file
+	if !needPackage && bctx.CgoEnabled {
+		return match, false, "", nil
+	}
+	// if it's not a go file, there is no package or cgo
+	if !strings.HasSuffix(input, ".go") {
+		return match, false, "", nil
+	}
+	data, err := ioutil.ReadFile(input)
+	if err != nil {
+		return false, false, "", err
+	}
+	isCgo, pkg, err := testCgo(input, data)
+	if err != nil {
+		return false, false, "", err
+	}
+	// match if cgo is enabled or the file is not cgo
+	return match && (bctx.CgoEnabled || !isCgo), isCgo, pkg, err
+}
+
+func testCgo(src string, data []byte) (bool, string, error) {
```

Comment:
Instead of requiring callers to pass in the full contents, let's just pass the filename to `ParseFile`. This should let us avoid reading the entire file, and will simplify calling code.

Also, name return values.

Also, since this more generally collecting metadata now, maybe rename to something like `readGoMetadata`?

Sorry I know this function was just moved verbatim. Feel free to tell me to do it in another PR. :)